### DISCUSSION TITLE
Reverting to latest Chromedriver version when provided version not available

### DISCUSSION THREAD

**Discussion by jdholtz** [seleniumbase/SeleniumBase#2244] [Q&A] [answered]
Hello. I am using the new `driver_version="browser"` parameter for the UC driver. It works great, but currently there seems to not be a [Chromedriver version for 119.0.6045.106](https://googlechromelabs.github.io/chrome-for-testing/) even though (for some reason) there is a [.106 Chrome version for Windows](https://chromereleases.googleblog.com/2023/10/stable-channel-update-for-desktop_31.html) which is causing the driver to fail to download.

Would there be a way, using SeleniumBase, to detect that the current version is not available yet and revert back to the latest driver version? The exception thrown right now ([line 553 of sb_install.py](https://github.com/seleniumbase/SeleniumBase/blob/466f8d4a25d6c948f40d68b36bac27d46d25d245/seleniumbase/console_scripts/sb_install.py#L553)) is very general so there isn't a good way to ensure it is only because the specific version isn't found.

A possible solution is to also compare the browser version to the latest version and revert to the latest version if the browser version is bigger than the driver version. I'm not familiar with the internals of SB though, so I'm not sure if this is optimal or not.

I am currently using SB 4.20.9 on Python 3.11.5.

**Comment by mdmintz** [comment, marked as answer]
Hello. Instead of using `driver_version="browser"`, use `driver_version="mlatest"`, as described here: https://github.com/seleniumbase/SeleniumBase/issues/2036#issue-1868891198.

That option specifically gives you the latest milestone version of chromedriver that matches your milestone version of Chrome. These drivers are built in such a way that the latest driver for a milestone will work with any browser of the same milestone. (Eg. `chromedriver` 118 for Chrome 118.) It's an upgrade over matching the exact version of the driver to the browser, because if there's a bug with a driver version (but not the matching exact browser version), then they could toss the driver version and replace it with a newer one without necessarily replacing the browser version.

And newer drivers of the same milestone are built to be fully backwards compatible with any browser of the same milestone version.

Therefore, use `--driver-version=mlatest` / `driver_version="mlatest"`. There haven't been any reported issues with that (so far).
I would need to see at least one issue that I can reproduce with that before changing the system around again.

**Reply by jdholtz** [reply]
That does work well. However, it seems to download the driver every time I use it. Is it possible to have logic instead that only downloads the `mlatest` if the browser versions do no match? For example, let's say I have browser version 119.0.0.1. If I have driver version 119.0.0.1 from a previous download, it wouldn't attempt to download a new one since they are compatible. However, if I have driver 119.0.0.0, it would try to download `mlatest`, being 119.0.0.1. 

The reason I want to try to avoid downloading it is because the only valid Chromedriver for Linux ARM is from the official repositories, but the Chromedriver team doesn't provide a specific one for Linux ARM, so SB ends up installing the x86 one which results in an exec error. Therefore, I currently move the Chromedriver downloaded from the official repository into the correct place to "trick" SB into thinking the driver and browser version match so it doesn't download a new one.

**Reply by mdmintz** [reply]
I see. `driver = Driver(driver_version="mlatest")` works.
But `driver = Driver(driver_version="mlatest", uc=True)` keeps downloading it again.
To quote Batman, "I'll look into it."

**Reply by jdholtz** [reply]
If this helps, the reason why it is triggering a download is due to lines [2989-2993 in core/sb_install.py](https://github.com/seleniumbase/SeleniumBase/blob/466f8d4a25d6c948f40d68b36bac27d46d25d245/seleniumbase/core/browser_launcher.py#L2989).

**Reply by mdmintz** [reply]
Thanks! I'll have a fix for that in the next release.

**Reply by mdmintz** [reply]
Fixed in `4.21.0`, now available: https://github.com/seleniumbase/SeleniumBase/releases/tag/v4.21.0

**Reply by jdholtz** [reply]
Thanks! It works perfectly now.

> Otherwise, if the major browser version matches the major driver version (eg. 117.* and 117.*), then no new driver is downloaded, even if the driver version is not the latest version for that milestone.

This is from the original comment you linked to. On the Python 3.12 Alpine image, the latest version of both Chromium and the chromedriver from the APK repositories is 117.0.5938.62. However, with `mlatest`, it attempts to install the newest version in the same major version. This is expected behavior, right? The quote above makes me think the new one should not be installed if the browser and driver versions match.

**Reply by mdmintz** [reply]
That's expected. `driver_version="mlatest"` will make sure your driver is the latest one available for that milestone.

If you want to keep your existing driver, use `driver_version="keep"` (as long as you have any driver at all, even if it's a completely different milestone version, then it will keep using it).
